### PR TITLE
Collect vpp postmortem data (#738)

### PR DESCRIPTION
**Pull request by stepan-anokhin**
## Description
Automates vpp-dataplane postmortem data collection.  

If `networkservicemesh/vppagent-dataplane` is run with `-e COLLECT_POSTMORTEM_DATA=true` it will store vpp core dump, api traces and logs at `$POSTMORTEM_DATA_LOCATION` (default is `/var/tmp/nsm-postmortem/vpp-dataplane`)

## Motivation and Context
See #738 

## How Has This Been Tested?
- [ ] Covered by existing integration testing
- [x] Added integration testing to cover
- [x] Tested locally

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: dataplane/vppagent/conf/vpp/startup.conf [hunk 1]
```diff
@@ -71,7 +71,7 @@ cpu {
 }
 
 # dpdk {
-	## Change default settings for all intefaces
+	## Change default settings for all interfaces
```
Comment (edwarnicke) [inline dataplane/vppagent/conf/vpp/startup.conf]: LOL... I appreciate your fastidious attention to detail :)

**Comment by edwarnicke** [comment]
This looks good so far... what are the plans for extracting the postmortem data to the CI jobs?

File: dataplane/vppagent/conf/vpp/startup.conf (hunk 1 above)
Comment (stepan-anokhin) [reply dataplane/vppagent/conf/vpp/startup.conf]: Hahahahah :)) Why not?

**Comment by haiodo** [comment]
@stepan-anokhin Please sign your commits.

File: dataplane/vppagent/scripts/vpp_monitor_handlers.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    gdb.execute("cd /tmp")
+    gdb.execute("set logging file vpp_backtrace")
+    gdb.execute("set logging on")
+    gdb.execute("thread apply all bt")
+    gdb.execute("set logging off")
+
+
+def save_core_dump():
+    gdb.execute("cd " + core_dump_location())
+    gdb.execute("generate-core-file")
```
Comment (haiodo) [inline dataplane/vppagent/scripts/vpp_monitor_handlers.py]: Could we compress core files using gzip2? Should they be smaller in this case?

Comment (edwarnicke) [reply dataplane/vppagent/scripts/vpp_monitor_handlers.py]: Good idea :)  Corefiles are *huge*

File: dataplane/vppagent/conf/supervisord/supervisord-dev.conf [hunk 3]
```diff
@@ -29,7 +38,7 @@ command=/bin/vppagent-dataplane
 startsecs=0
 autorestart=false
 redirect_stderr=true
-priority=3
+priority=4
```
Comment (edwarnicke) [inline dataplane/vppagent/conf/supervisord/supervisord-dev.conf]: Curious... why priority four here?

Comment (stepan-anokhin) [reply dataplane/vppagent/conf/supervisord/supervisord-dev.conf]: Ah, this one actually doesn't play a role. 

Historically I'was trying to resolve one issue related to health-probing via sub-process prioritization.
When GDB attaches to vpp, the vpp usually freezes for about 0.7s. This is enough to cause intermittent issue with agent complaining on non-responsive vpp and exiting with error. I've [increased](https://github.com/networkservicemesh/networkservicemesh/pull/762/files/c0c9a6c7e5eb532d892a566c2d35a5875d70c0b5#diff-d9a7275054fd2ab8910d32d4a5f0762aR28) reply timeout in dev-image to resolve this issue, but forgot to rollback changes in `supervisord-dev.conf`. 

Currently it shouldn't affect anything.

**Comment by stepan-anokhin** [comment]
Build failed with 
```
/usr/local/go/pkg/tool/linux_amd64/link: signal: killed
Exited with code 1
```

That seems to be intermittent problem unrelated to my changes. I don't have permissions to restart the build. :(

**Comment by stepan-anokhin** [comment]
Only VPN tests fail. vpp stack-trace was successfully collected BTW.

**Comment by edwarnicke** [comment]
I think you may be making this to complex.

To get the core dumps... you basically just need to add the bash shell incantations to set up for core dumps [here](https://github.com/networkservicemesh/networkservicemesh/blob/master/.mk/packet.mk#L34) in packet-start after the ./create-kubernetes-cluster.sh script.  You can write a new simple script [like this](https://github.com/networkservicemesh/networkservicemesh/blob/master/scripts/vagrant/scripts/coredump.sh) (or just symlink it into scripts/terraform).

Then you can just recover them in the .circleci/config.yml with scp from the packet boxes to the directory you setup to retrieve them.

**Comment by stepan-anokhin** [comment]
@edwarnicke 
resetting `kernel.core_pattern` affects host OS and all the rest of containers on the same node. Initially it was done exactly like that, but than I switched to gdb to overcome this problem.

**Comment by edwarnicke** [comment]
@stepan-anokhin For capturing these failures in CI its fine to capture at the host level.  Keep it simple :)

**Comment by nickolaev** [comment]
#832 merged and obsoletes this one. Closing. Please reopen if needed.

**Pull request closed without merging**